<a href="https://colab.research.google.com/github/StacyChebet/LoanDefaults/blob/master/LoanDefaults.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Introduction to Data Processing**
Data processing involves cleaning and transforming raw data into a suitable format for modeling.
## **Objective**
This notebook aims to provide hands-on experience in preprocessing data, focusing on handling missing value, scaling feautures, encoding categorical variables, and more.
## **Key Objectives**
1. **Understand Data:** Familiarize with the dataset and its characteristics.
2. **Handle Missing Values:** Identify and treat missing data in different columns.
3. **Feature Encoding:** Convert categorical variables into numerical formats suitable for machine learning models.
4. **Feature Scaling:** Normalize or standardize numerical values to improve model performance.
5. **Data Splitting:** Prepare the dataset for training and testing to evaluate the performance of machine learning models.


##**Data Description**
The dataset provided is a collection of loan application records, which can be used to ptedict the likelihood of a default. <br>
Here is a brief description of each column in the dataset:
- `TARGET`: Binary indicator where **1** represents a default on a loan and **0** represents a non-default. This is the label for our predictive modeling.
- `NAME_CONTRACT_TYPE`: Type of loan contracted. Categorical variable (e.g. 'Cash loans', 'Resolving loans').
- `CODE_GENDER`: Gender of the applicant. Categorical variable ('M' for male, 'F' for female).
- `FLAG_OWN_CAR`: Indicates whether the applicant owns a car ('Y' for yes, 'N' for no).
- `FLAG_OWN_REALTY`: Indicates whether the applicant owns real estate ('Y' for yes, 'N' for no).
- `CNT_CHILDREN`: Number of children the applicant has.
- `AMT_INCOME_TOTAL`: Total annual income of the applicant.
- `AMT_CREDIT`: Credit amount of the loan taken.
- `AMT_ANNUITY`: Loan annuity.
- `DAYS_BIRTH`: Applicant's age in days at the time of application (negative values indicating the age).
- `YEARS_EMPLOYED`: Number of years the applicant has been employed.


##**Loading Libraries and Data**
Libraries used:
- **Pandas:** For data manipulation
- **Numpy:** For numerical operations
- **Seaborn:** For data visualization

In [1]:
#Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#Setting visualization styles
sns.set(style="whitegrid")

#Mounting google drive
from google.colab import drive
drive.mount('/content/drive')

#Changing directory
%cd /content/drive/My Drive/Colab Notebooks/Data Analytics - IBT/LoanDefaults

#Loading the dataset
file_path = "loan_default.csv"
df = pd.read_csv(file_path)

#Displaying the first few rows of the dataset
df.head()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
/content/drive/My Drive/Colab Notebooks/Data Analytics - IBT/LoanDefaults


TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  \
0     0.0         Cash loans           M            Y               N   
1     0.0    Revolving loans           M            Y               Y   
2     0.0         Cash loans           M            Y               N   
3     0.0         Cash loans           F            N               Y   
4     0.0         Cash loans           F            N               Y   

   CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  DAYS_BIRTH  \
0             1          225000.0    578619.0      23229.0      -12347   
1             1               NaN    270000.0      13500.0      -14048   
2             0          144000.0    753840.0      29340.0      -14639   
3             0           81000.0     98910.0       7785.0      -14591   
4             1          103500.0    521280.0      26779.5      -12023   

   YEARS_EMPLOYED  
0               0  
1               6  
2               6  
3              11  
4               0

## **Understanding the Dataset**

In this section, we'll dive into understanding some crucial aspects of our dataset. This foundational knowledge is key to effective preprocessing and ultimately, to building robust machine learning models. Here are the areas we will focus on:

- **Data Shape and Structure**: Understanding the size of the dataset and the structure of the data tables. It’s important to know how many instances (rows) and features (columns) we have.
  
- **Summary Statistics**: By examining the central tendency and spread of numeric features, we can gain insights into the general distribution and scale of our data.

- **Data Types**: Different data types require different preprocessing methods. We will identify which columns are numerical, categorical, or boolean to tailor our preprocessing techniques appropriately.

- **Missing Values**: Identifying missing values is crucial as they can significantly impact the performance of machine learning models. We'll need to decide how to handle these, whether by imputation, deletion, or some other method.

- **Unique Values and Cardinality**: Understanding the uniqueness of data, especially for categorical data, helps in determining how to handle encoding and can reveal insights about the dataset’s granularity.

- **Potential Errors or Outliers**: Early detection of anomalies or outliers can help us understand data collection errors or exceptional cases, which may require special handling or exclusion to avoid skewing our model results.

- **Initial Observations**: We'll note any initial observations that may require deeper analysis or might impact how we choose to preprocess the data.

Understanding these elements will set the stage for the data cleaning and preparation tasks that follow, ensuring that we approach the preprocessing with a comprehensive understanding of our dataset’s characteristics.

##**Initial Data Exploration**
In this step, we will conduct an initial exploration of the dataset to understand its structure and basic characteristics. We will:

1. Check the shape of the dataset
2. Display the data types of each column
3. Get a summary of the dataset using descriptive statistics

In [2]:
#Checking the shape of the dataset
print(f"The dataset contains {df.shape[0]} rows and {df.shape[1]} columns.")

#Displaying the data types of each column
print("\nData types of each column:")
print(df.dtypes)

#Getting a summary of the dataset using descriptive statistics
print("\nSummary of the dataset:")
print(df.describe().T)

The dataset contains 10284 rows and 11 columns.

Data types of each column:
TARGET                float64
NAME_CONTRACT_TYPE     object
CODE_GENDER            object
FLAG_OWN_CAR           object
FLAG_OWN_REALTY        object
CNT_CHILDREN            int64
AMT_INCOME_TOTAL      float64
AMT_CREDIT            float64
AMT_ANNUITY           float64
DAYS_BIRTH              int64
YEARS_EMPLOYED          int64
dtype: object

Summary of the dataset:
                    count           mean            std      min         25%  \
TARGET            10281.0       0.079078       0.269873      0.0       0.000   
CNT_CHILDREN      10284.0      10.287340     986.138233      0.0       0.000   
AMT_INCOME_TOTAL   6993.0  167909.947902  105245.713874  27000.0  112500.000   
AMT_CREDIT        10284.0  599902.578326  400843.758473  45000.0  270000.000   
AMT_ANNUITY       10284.0   27191.230747   14933.326225   2974.5   16509.375   
DAYS_BIRTH        10284.0  -16055.232400    4377.738214 -25166.0  -19724.25

##**Findings**
1. **Dataset Overview:**
- The dataset contains 10284 rows and 11 columns.
- Features include both **numerical** (e.g. `TARGET`, `DAYS_BIRTH`) and **categorical** (e.g. `NAME_CONTRACT_TYPE`, `DAYS_BIRTH`) variables.

2. **Key Insights**
- `AMT_INCOME_TOTAL`: The maximum income amount is 3,825,000 and the minimum amount is 27,000.<br>
 - The count of this variable is significantly less than the count of other variables, showing that there are missing values (3291 missing values).
- `YEARS_EMPLOYED`: The maximum amount of years an applicant has been employed is 150 years, and the minimum is -1000 years, both of which are far-fetched, which might mean there are errors.
- `CNT_CHILDREN`: The maximum number of children is 99,999 which is an impossible figure.


##**Summary Statistics for Numeric Columns**
- `TARGET`: The mean of approximately 0.08 suggests that about 8% of the entries are defaults.
- `CNT_CHILDREN`: The count of children ranges widely from 0 to 99,999, with a mean close to 10. This unusually high maximum might indicate an error or outlier.
- `AMT_INCOME_TOTAL`: Total income ranges from 27,000 to 3,825,000, with a significant standard deviation, suggesting wide disparity in applicants' income levels.
- `AMT_CREDIT`: The loan credit amount varies from 45,000 to 4,050,000, highlighting the broad range of loan amounts processed.
- `AMT_ANNUITY`: Annuities range from 2,974.5 to 258,025.5, showing what borrowers pay periodically, with a mean of about 27,191.
- `DAYS_BIRTH`: The days since birth (negative values indicating age) range from -25,166 to -7,680, which corresponds to ages approximately between 20 and 69 years.
- `YEARS_EMPLOYED`: Employment duration ranges from -1000 to 150 years, where negative values may indicate errors or specific data entry conventions needing clarification.

## **Checking for Missing Values**

In [3]:
#Check for missing values in each column
print("\nMissing Values in each column:")
print(df.isnull().sum())


Missing Values in each column:
TARGET                   3
NAME_CONTRACT_TYPE       0
CODE_GENDER              0
FLAG_OWN_CAR          2262
FLAG_OWN_REALTY         30
CNT_CHILDREN             0
AMT_INCOME_TOTAL      3291
AMT_CREDIT               0
AMT_ANNUITY              0
DAYS_BIRTH               0
YEARS_EMPLOYED           0
dtype: int64


##**Missing Values Analysis**
Missing data can significantly affect the outcomes of our analysis and predictive modeling. <br>
Here's a breakdown of missing values across different columns in the dataset:
- `TARGET`: 3 missing values. This is crucial since it is our dependent variable. We need to decide whether to impute these values or remove the corresponding records.
- `NAME_CONTRACT_TYPE`, `CODE_GENDER`, `CNT_CHILDREN`, `AMT_CREDIT`, `AMT_ANNUITY`, `DAYS_BIRTH`, `YEARS_EMPLOYED`: No missing values, which simplifies preprocessing for these columns.
- `FLAG_OWN_CAR`: 2,262 missing values. We need to determine whether this is due to data collection process or other reasons, and decide on an appropriate imputation strategy.
- `FLAG_OWN_REALTY`: 30 missing values. Considering the small number relative to the dataset size, we might opt for simple imputation methods like mode replacement.
- `AMT_INCOME_TOTAL`: 3,291 missing values, a significant number that requires careful consideration for imputation, as it could impact any analysis related to the applicant's income.

#**Check for Unique Values to Understand Cardinality**


In [4]:
#Check the number of unique values in each column to understand cardinality
print("\nUnique Values in Each Column:")
print(df.nunique())


Unique Values in Each Column:
TARGET                   2
NAME_CONTRACT_TYPE       2
CODE_GENDER              2
FLAG_OWN_CAR             2
FLAG_OWN_REALTY          4
CNT_CHILDREN            13
AMT_INCOME_TOTAL       260
AMT_CREDIT            1860
AMT_ANNUITY           4279
DAYS_BIRTH            7604
YEARS_EMPLOYED          46
dtype: int64


##**Cardinality Analysis**
**Cardinality** refers to the number of unique values in a column. <br>
Understanding cardinality is important for identifying features with a wide range of values and those with limited variety, which influences how we might process and use these features in machine learning models. <br>
Below is a summary of the cardinality in each column of our dataset:
- `TARGET`: 2 unique values. This binary variable indicates whether a loan defaulted or not, suitable for classification tasks.
- `NAME_CONTRACT_TYPE`: 2 unique values. Indicates the type of loan, either 'Cash loans' or 'Revolving loans'.
- `CODE_GENDER`: 2 unique values, typically 'M' for male and 'F' for female.
- `FLAG_OWN_CAR`: 2 unique values, indicating car ownership ('Y' for yes and 'N' for no).
- `FLAG_OWN_REALTY`: 4 unique values, suggesting some categories or errors since typically this should be 'Y' or 'N'.
- `CNT_CHILDREN`: 13 unique values, showing the number of children ranging from 0 to a higher number, affecting the applicant's financial responsibilities.
- `AMT_INCOME_TOTAL`: 260 unique values. The income amounts vary significantly, which could reflect different economic statuses.
- `AMT_CREDIT`: 1,860 unique values, showing a wide range of loan amounts.
- `AMT_ANNUITY`: 4,279 unique values, indicating varied periodic repayment plans.
- `DAYS_BIRTH`: 7,604 unique values, translating directly to the age of the clients.
- `YEARS_EMPLOYED`: 46 unique values, showing diverese employment durations among applicants.

The high cardinality in columns like `AMT_CREDIT`, `AMT_ANNUITY` and `DAYS_BIRTH` indicates **continuos numerical data**, which may require **binning** or **scaling** during preprocessing.<br>
Conversely, low cardinality columns like `TARGET` or `CODE_GENDER` might be more straightforward to handle but are crucial for any analytical or predictive modeling process.

In [5]:
#Checking the unique values in `FLAG_OWN_REALTY`
df.FLAG_OWN_REALTY.unique()

array(['N', 'Y', nan, 'YES', 'NO'], dtype=object)

##**Treating Inconsistencies in `FLAG_OWN_REALTY`**
The values in this column should either be a straightforward 'Yes' or 'No' ('Y' or 'N'). <br>
However, a closer inspection reveals inconsistencies in how the data is entered, with variations such as 'Y', 'N', 'YES', 'NO', and missing values.<br>
Standardizing this column is crucial for accurate analysis and modeling.<br>
Here are the speps we'll take: <br>
**Standardization:** Convert all variations to a uniform format ('Y' and 'N').

In [6]:
#Mapping the values to standard form
df['FLAG_OWN_REALTY'] = df['FLAG_OWN_REALTY'].replace({'YES':'Y', 'NO':'N'})

df.FLAG_OWN_REALTY.unique()

array(['N', 'Y', nan], dtype=object)

##**Treating Missing Values in the Dataset**
Missing values can introduce bias and affect the performance of machine learning models.<br>
The following columns have missing values that need to be addressed before further analysis:
- `TARGET` Column
This dependent variable, has 3 missing entries. Given the small number of missing values relative to the dataset size, and the importance of having complete labels for all training instances in supervised learning, the most appropriate action is to drop these rows.
- `FLAG_OWN_CAR` Column
This column contains 2,262 missing entries. Since this is a categorical variable, we will impute the missing values using the **mode** of this column, representing the most common category.
- `AMT_INCOME_TOTAL` Column
This column has 3,291 missing values. We will use the **median** to impute these values, which is preferred over the mean because it is more **robust to outliers**, which are evident from the wide range of income values present in our dataset.

In [8]:
#Drop rows where `TARGET` is missing
df.dropna(subset=['TARGET'], inplace=True)

#Impute missing value in `FLAG_OWN_REALTY` with mode
realty_mode = df['FLAG_OWN_REALTY'].mode()[0]
df['FLAG_OWN_REALTY'] = df['FLAG_OWN_REALTY'].fillna(realty_mode)

#Impute missing values in `FLAG_OWN_CAR` with mode
car_mode = df['FLAG_OWN_CAR'].mode()[0]
df['FLAG_OWN_CAR'] = df['FLAG_OWN_CAR'].fillna(car_mode)

#Impute missing values in `AMT_INCOME_TOTAL` with median
income_median = df['AMT_INCOME_TOTAL'].median()
df['AMT_INCOME_TOTAL'] = df['AMT_INCOME_TOTAL'].fillna(income_median)

#Display the summary to verify changes
print(df.isnull().sum())

TARGET                0
NAME_CONTRACT_TYPE    0
CODE_GENDER           0
FLAG_OWN_CAR          0
FLAG_OWN_REALTY       0
CNT_CHILDREN          0
AMT_INCOME_TOTAL      0
AMT_CREDIT            0
AMT_ANNUITY           0
DAYS_BIRTH            0
YEARS_EMPLOYED        0
dtype: int64
